indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java

```diff
@@ -123,7 +135,16 @@ private static String makeDatasource(FireDepartment fireDepartment)
   private final FireDepartment spec;
 
   @JsonIgnore
-  private volatile Plumber plumber = null;
+  private final Queue<ListenableFuture<SegmentsAndMetadata>> pendingPublishes;
```

Comment:
I originally had it that way but it introduces a race condition when the user specifies a handoff timeout.

When the user specifies a handoff timeout, its possible that the handoff timeout expires before the segment is published.  That was why `RealtimeIndexTask#testHandoffTimeout` was intermittently failing on travis.

If this isn't a concern, I can refactor that away and modify the test.

Reply:
Race conditions should be fixed, but I'm not sure where it can exist. Would you elaborate more on this?
What I first thought was keeping all segments in the pendingHandoff state like below.

```java
    pendingHandoffs.add(
        Futures.transform(
            driver.publish(publisher, committerSupplier.get(), Collections.singletonList(sequenceName)),
            (AsyncFunction<SegmentsAndMetadata, SegmentsAndMetadata>) segmentsAndMetadata -> driver.registerHandoff(segmentsAndMetadata)
        )
    );
```

Reply:
The race condition was in `RealtimeIndexTask#testHandoffTimeout`.  That test expects the handoff to timeout but still have the segments be published.

If you wait on the combined (publish + handoff) future with a timeout, its possible that the timeout expires before the segment is published